packages/@aws-cdk/core/lib/private/metadata-resource.ts

```diff
@@ -76,17 +47,83 @@ function makeCdkMetadataAvailableCondition() {
     .map(ri => Fn.conditionEquals(Aws.REGION, ri.name)));
 }
 
-function formatModules(runtime: cxapi.RuntimeInfo): string {
-  const modules = new Array<string>();
+/** Convenience type for arbitrarily-nested map */
+class Trie extends Map<string, Trie> { }
 
-  // inject toolkit version to list of modules
-  const cliVersion = process.env[cxapi.CLI_VERSION_ENV];
-  if (cliVersion) {
-    modules.push(`aws-cdk=${cliVersion}`);
-  }
+/**
+ * Formats a list of construct fully-qualified names (FQNs) and versions into a (possibly compressed) prefix-encoded string.
+ *
+ * The list of ConstructInfos is logically formatted into:
+ * ${version}!${fqn} (e.g., "1.90.0!aws-cdk-lib.Stack")
+ * and then all of the construct-versions are grouped with common prefixes together, grouping common parts in '{}' and separating items with ','.
+ *
+ * Example:
+ * [1.90.0!aws-cdk-lib.Stack, 1.90.0!aws-cdk-lib.Construct, 1.90.0!aws-cdk-lib.service.Resource, 0.42.1!aws-cdk-lib-experiments.NewStuff]
+ * Becomes:
+ * 1.90.0!aws-cdk-lib.{Stack,Construct,service.Resource},0.42.1!aws-cdk-lib-experiments.NewStuff
+ *
+ * The whole thing is then either included directly as plaintext as:
+ * v2:plaintext:{prefixEncodedList}
+ * Or is compressed and base64-encoded, and then formatted as:
+ * v2:deflate64:{prefixEncodedListCompressedAndEncoded}
+ *
+ * Exported/visible (and `forcePlaintext` parameter) for ease of testing.
+ */
+export function formatAnalytics(infos: ConstructInfo[], forcePlaintext: boolean = false) {
+  const trie = new Trie();
+  infos.forEach(info => insertFqnInTrie(`${info.version}!${info.fqn}`, trie));
 
-  for (const key of Object.keys(runtime.libraries).sort()) {
-    modules.push(`${key}=${runtime.libraries[key]}`);
-  }
-  return modules.join(',');
-}
\ No newline at end of file
+  const plaintextEncodedConstructs = prefixEncodeTrie(trie);
+  const compressedConstructs = zlib.gzipSync(Buffer.from(plaintextEncodedConstructs)).toString('base64');
+
+  return (plaintextEncodedConstructs.length < compressedConstructs.length || forcePlaintext)
```

Comment:
feels like a premature optimization. do we really need to support plaintext? (for debugging we can just emit this to STDERR when --verbose is set)

Reply:
Totally fair. For really simple apps, it's nice to be able to see the output, but anything past ~12-15 unique constructs crosses over the threshold.

Reply:
So did you decide to leave it like this or simplify?